### PR TITLE
Governance initializers

### PR DESCRIPTION
**Pull request by alpe**
* Genesis initializer for Electorate and Election rules
* model validation rules proposal

See #439

### CODE REVIEW

File: x/gov/model.go [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+import (
+	"fmt"
+	"regexp"
+
+	"github.com/iov-one/weave"
+	"github.com/iov-one/weave/errors"
+	"github.com/iov-one/weave/orm"
+)
+
+var validTitle = regexp.MustCompile(`^[a-zA-Z0-9_.-]{4,128}$`).MatchString
```
Comment (ethanfrey) [inline x/gov/model.go]: Allow space in title as well

File: x/gov/model.go [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+func (m Electorate) Copy() orm.CloneableData {
+	p := make([]Participant, 0, len(m.Participants))
+	copy(p, m.Participants)
+	return &Electorate{
+		Title:        m.Title,
+		Participants: p,
+	}
+}
+
+const maxWeight = 2 ^ 16 - 1
```
Comment (ethanfrey) [inline x/gov/model.go]: Fair enough

File: x/gov/model.go [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+	rev, ok := obj.Value().(*Electorate)
+	if !ok {
+		return nil, errors.Wrapf(errors.ErrInvalidModel, "invalid type: %T", obj.Value())
+	}
+	return rev, nil
+}
+
+const (
+	minVotingPeriodHours = 1
+	maxVotingPeriodHours = 365 * 24
```
Comment (ethanfrey) [inline x/gov/model.go]: Ooof.. That is huge. How about limiting it to 4 weeks... 28 days.

File: x/gov/model.go [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+	return &ElectionRule{
+		Title:             m.Title,
+		VotingPeriodHours: m.VotingPeriodHours,
+		Threshold:         m.Threshold,
+	}
+}
+
+func (m Fraction) Validate() error {
+	switch {
+	case m.Numerator == 0:
```
Comment (ethanfrey) [inline x/gov/model.go]: I think 0 is fine.
For me fraction is a number between [0, 1] inclusive, right?
Unless we use it only for voting power, when it must be between (0.5, 1]

I would either change this to `m.Numerator < 0` (first case) or `m.Numerator *2 <= m.Denominator` (second case) and then document that on top of this function

File: x/gov/model.go [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+		return errors.Wrap(errors.ErrInvalidInput, "numerator must not be 0")
+	case m.Denominator == 0:
```
Comment (ethanfrey) [inline x/gov/model.go]: `m.Denominator <= 0` should be prohibited. Let's keep out negatives for our sanity

Comment (alpe) [reply x/gov/model.go]: good point but it is uint32, so no negative values are possible

File: x/gov/model.go (hunk 4 above)
Comment (alpe) [reply x/gov/model.go]: The uint32 type prevents negative values. I would like to keep it in the [0.5,1] range as the type is used by the gov module only. I will document the threshold better. Thanks, good input!

**Pull request merged**